### PR TITLE
CDAP-3902: Remove unnecessary remote calls when getting a dataset instance

### PR DESCRIPTION
**Pull request by jwang47**
- Added DatasetProvider and DatasetMetaProvider
- Made UsageRegistry obtain the usage registry dataset instance by using
  DatasetProvider instead of DatasetFramework
- Removed namespace existence check in DatasetInstanceService.get()

https://issues.cask.co/browse/CDAP-3902
http://builds.cask.co/browse/CDAP-DUT2984

Reopened from https://github.com/caskdata/cdap/pull/4320

### After changes (total time taken)
N=300: 5787, 6596, 6201, 7068, 6513
N=500: 7702, 7985, 7871, 7920, 9149
N=1000: 12762, 13178, 13163, 12120
N=2000: 19436 (w/ a few ServiceUnavailableExceptions), 15377 (w/ a few ServiceUnavailableExceptions)

### Before changes (total time taken)
N=300: 15535, 11385, 9931, 9153, 8549
N=500: 18091 (w/ timeouts), 17249 (w/ timeouts)

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    DatasetProperties creationProps,
+    @Nullable ClassLoader classLoader,
+    @Nullable Map<String, String> arguments) throws Exception {
+
+    T result = get(instance, classLoader, arguments);
+    if (result != null) {
+      return result;
+    }
+
+    return null;
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java]: This should call `createIfNotExists`, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Provides {@link Dataset} instances using methods implemented by subclasses.
+ * Used by {@link RemoteDatasetFramework} and {@link DatasetInstanceService}.
+ * Use this when you want to control how dataset instances are created. For example, when
+ * you want to obtain a {@link Dataset} instance without having to make remote calls.
+ */
+public abstract class LocalDatasetProvider implements DatasetProvider {
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java]: Maybe call it `AbstractLocalDatasetProvider`, although it is a bit long. What do you think?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java [hunk 3]
```diff
@@ -65,12 +70,28 @@
   @Inject
   public DatasetInstanceService(DatasetTypeManager implManager, DatasetInstanceManager instanceManager,
                                 DatasetOpExecutor opExecutorClient, ExploreFacade exploreFacade, CConfiguration conf,
-                                UsageRegistry usageRegistry, AbstractNamespaceClient namespaceClient) {
+                                TransactionExecutorFactory txFactory,
+                                DatasetDefinitionRegistryFactory registryFactory,
+                                AbstractNamespaceClient namespaceClient,
+                                DatasetFramework framework) {
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: `DatasetFramework` should no longer be necessary now, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java [hunk 4]
```diff
@@ -117,6 +136,10 @@ public DatasetMeta get(Id.DatasetInstance instance, List<? extends Id> owners) t
     return new DatasetMeta(spec, typeMeta, null);
   }
 
+  public DatasetMeta getMeta(Id.DatasetInstance instance) throws Exception {
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: This can be in-lined so as to not add another public method.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java (hunk 2 above)
Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java]: How about AbstractDatasetProvider since it's not necessarily local?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 5]
```diff
@@ -160,6 +168,11 @@ public DatasetSpecification getDatasetSpec(Id.DatasetInstance datasetInstanceId)
     return meta == null ? null : meta.getSpec();
   }
 
+  public DatasetMeta getMeta(Id.DatasetInstance instance) throws Exception {
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: This can be in-lined so as to not add another public method.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java [hunk 6]
```diff
@@ -88,7 +81,22 @@ public DatasetServiceClient load(Id.Namespace namespace) throws Exception {
...
+      }
+
+      @Override
+      public void createIfNotExists(Id.DatasetInstance instance, String type,
+                                    DatasetProperties creationProps) throws Exception {
+        if (getMeta(instance) != null) {
+          return;
+        }
+
+        RemoteDatasetFramework.this.addInstance(type, instance, creationProps);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: Won't this throw exception if instance already exists?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java [hunk 7]
```diff
@@ -171,6 +194,15 @@ public void create(Id.Namespace namespace, String name, DatasetInstanceConfigura
     enableExplore(newInstance, props);
   }
 
+  public void createIfNotExists(Id.Namespace namespace, String name,
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: This can be a private method, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/UsageRegistry.java [hunk 8]
```diff
@@ -52,14 +54,22 @@
...
       @Override
       public UsageDatasetIterable get() {
         try {
-          Object usageDataset = DatasetsUtil.getOrCreateDataset(datasetFramework, USAGE_INSTANCE_ID,
-                                                                UsageDataset.class.getSimpleName(),
-                                                                DatasetProperties.EMPTY, null, null);
+          Object usageDataset = provider.get(USAGE_INSTANCE_ID, null, null);
+          if (usageDataset == null) {
+            // use DatasetFramework only to create
+            DatasetsUtil.createIfNotExists(framework, USAGE_INSTANCE_ID,
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/UsageRegistry.java]: This is no longer required, right?

**Comment by poornachandra** [comment]
@jwang47 Please address comments.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java (hunk 6 above)
Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: Yes, but if the instance already exists, this line won't get executed due to the previous if-statement.

Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: What about race conditions?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+public abstract class AbstractDatasetProvider implements DatasetProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractDatasetProvider.class);
+  private final DatasetDefinitionRegistryFactory registryFactory;
+
+  protected AbstractDatasetProvider(DatasetDefinitionRegistryFactory registryFactory) {
+    this.registryFactory = registryFactory;
+  }
+
+  public abstract DatasetMeta getMeta(Id.DatasetInstance instance) throws Exception;
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java]: Need javadocs for abstract methods.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  public abstract void createIfNotExists(Id.DatasetInstance instance, String type,
+                                         DatasetProperties creationProps) throws Exception;
+
+  @Override
+  public <T extends Dataset> T getOrCreate(
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java]: Also, javadocs for public methods.

Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java]: Or, you could add javadocs at the interface level.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java (hunk 6 above)
Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/RemoteDatasetFramework.java]: Fixed

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java [hunk 11]
```diff
@@ -99,8 +118,6 @@ public DatasetInstanceService(DatasetTypeManager implManager, DatasetInstanceMan
    * @throws IOException if there is a problem in making an HTTP request to check if the namespace exists.
    */
   public DatasetMeta get(Id.DatasetInstance instance, List<? extends Id> owners) throws Exception {
-    // Throws NamespaceNotFoundException if the namespace does not exist
-    ensureNamespaceExists(instance.getNamespace());
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: If we are planning on fixing this in a later JIRA, then just comment it out and put the JIRA number here.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+    Id.DatasetInstance instance, String type,
+    DatasetProperties creationProps,
+    @Nullable ClassLoader classLoader,
+    @Nullable Map<String, String> arguments) throws Exception {
+
+    createIfNotExists(instance, type, creationProps);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java]: Doing a get first, and then a create only if get returns null will save on lots of create calls, right?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java [hunk 13]
```diff
@@ -171,6 +188,15 @@ public void create(Id.Namespace namespace, String name, DatasetInstanceConfigura
     enableExplore(newInstance, props);
   }
 
+  private void createIfNotExists(Id.Namespace namespace, String name,
+                                 DatasetInstanceConfiguration props) throws Exception {
+    try {
+      create(namespace, name, props);
```
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: It might be a good idea to do a check for existence before trying to create like you do in `RemoteDatasetFramework`.

Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: create() does this check.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+      // try program class loader then cdap class loader
+      try {
+        moduleClass = ClassLoaders.loadClass(moduleMeta.getClassName(), classLoader, this);
+      } catch (ClassNotFoundException e) {
+        try {
+          moduleClass = ClassLoaders.loadClass(moduleMeta.getClassName(), null, this);
+        } catch (ClassNotFoundException e2) {
+          LOG.error("Was not able to load dataset module class {} while trying to load type {}",
+                    moduleMeta.getClassName(), implementationInfo, e);
+          throw Throwables.propagate(e);
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java]: Do you think its a good idea to:
```
e.addSuppressed(e2);
```
before propagating it?

Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java]: Yes, but I'd prefer not to make any changes here in this PR. This PR moved this code here from somewhere else.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java (hunk 2 above)
Comment (poornachandra) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/LocalDatasetProvider.java]: Sounds good.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java (hunk 14 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java]: I think its fine to make the update the existing code, still, since its an improvement.

Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/AbstractDatasetProvider.java]: Fixed

File: cdap-master/src/main/java/co/cask/cdap/data/tools/UpgradeTool.java [hunk 15]
```diff
@@ -149,6 +151,21 @@ private Injector init() throws Exception {
       new LocationRuntimeModule().getDistributedModules(),
       new ZKClientModule(),
       new DiscoveryRuntimeModule().getDistributedModules(),
+      Modules.override(new DataSetsModules().getDistributedModules()).with(
```
Comment (poornachandra) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/UpgradeTool.java]: What is the reason for this change?

**Comment by poornachandra** [comment]
@jwang47 Please test it on a cluster by running integration tests. Also, let's get the performance numbers once more for the final set of changes.

File: cdap-master/src/main/java/co/cask/cdap/data/tools/UpgradeTool.java (hunk 15 above)
Comment (jwang47) [inline cdap-master/src/main/java/co/cask/cdap/data/tools/UpgradeTool.java]: UsageRegistry requires DatasetProvider, so I added DataSetsModules.
